### ISSUE TITLE
Index Checker: Expression in dependent type annotation invalid, identifier not found.

### ISSUE THREAD

**Issue by NhatDinh** [typetools/checker-framework#1900]
Hi, while using the index checker, I came across this problem.
Consider this simplified code block:
```java
public class SimpleMethod {
   void method1(@NonNegative @IndexFor("array") int m){
       @NonNegative @IndexFor("array") int n = m;
       int[] array = new int[n];
       array[n] = 0;
   }
}
```
```java
Error:(14, 50) java: [expression.unparsable.type.invalid] Expression in dependent type annotation invalid: [error for expression: array; error: 'array' because identifier not found]
```
 I think it is caused because array is declared after int n. 
I could not find documentation in the manual on how to deal with this problem. If it is a valid concern, I would like to know how we can fix it.

**Comment by panacekcz** [comment]
> I think it is caused because array is declared after int n.

Yes. The Index Checker gives this error for both the declaration of `n` and `m`. By declaring a variable of type `@IndexFor("array")`, you are requiring that the variable always contains a valid index for `array`. Such requirement does not have any meaning when `array` does not exist yet.

> I could not find documentation in the manual on how to deal with this problem.

It's hard to say what you are actually trying to achieve. The code always performs an out-of-bounds access, so there is no correct way to annotate it so that the Index Checker would not report any errors.

**Comment by NhatDinh** [comment]
Hi,
I'm sorry for not responding, I did not notice GitHub notification for this thread until now.
My concern for this issue is in the error message "required" type issued by Index Checker.
I have explained it clearer in [Issue #1913 ](https://github.com/typetools/checker-framework/issues/1913).
Please let me know what you think.
